### ISSUE TITLE
Tokens are still in the localStorage of application A when user logout from application B

### ISSUE THREAD

**Issue by boubou158** [okta/okta-auth-js#1387]
### Describe the bug?

I have a use case where a cookie is shared between:
- Application A on domain a.mydomain.com
- Application B on domain b.mydomain.com

When you open an application, the OICD flow is triggered if this cookie is present. The cookie is shared between those subdomains by being on the parent domain *.mydomain.com.

Let's assume the case:
- Login on application A with user Test1 => cookie is created
- Go to application B => cookie is here and no tokens are in the tokenManager so OICD flow is triggered and idToken/ accessToken are stored in the storage
- Logout on application A => cookie is deleted
- Login with another user Test2 on application A => cookie is created again
- Go to application B => cookie is here but since Test1 user tokens are still in the tokenManager, no new OICD call is done and logged in user displayed is still Test1 user.

How this is supposed to be handled?

Thanks in advance,

### What is expected to happen?

NA

### What is the actual behavior?

NA

### Reproduction Steps?

NA

### SDK Versions

NA

### Execution Environment

NA

**Comment by denysoblohin-okta** [comment]
Do you consider global Okta session (saved with cookie for domain `mydomain.com`) to be source of truth for current authorised user for both applications?

If yes, one of your options to manage your described case would be to compare user id from current access token saved for  app 2  with user id in [Okta session](https://developer.okta.com/docs/reference/api/sessions/#get-current-session). 
You can get user id from session with `(await oktaAuth.session.get()).userId` . 
And get user id from access token with `(await oktaAuth.tokenManager.get('accessToken')).claims.uid`
If user ids do not match, you can clear tokens for app 2 and authorize to get new tokens for user in Okta session.

If you want to completely separate applications and don't share Okta session, you can use [multiple custom domains](https://developer.okta.com/docs/guides/custom-url-domain/main/).

**Comment by boubou158** [comment]
Hi @denysoblohin-okta ,

Thanks for your reply. The problem with this solution is that session.get() is not third party cookies compatible. On chrome, i am able to see the current session, but I just tested it on firefox and the status will be INACTIVE due to third party blocking.

Do you have any other ideas in mind?

Thanks a lot in advance :)

**Comment by jaredperreault-okta** [comment]
Can you explain a bit more about your architecture? Do these applications share a centralized login experience? Or are they separate applications other than their "shared" Okta session?

**Comment by boubou158** [comment]
Hello @jaredperreault-okta ,

I will try to explain the current architecture:

All applications (A, B, C) are using the same standalone login page on the same parent domain, which is responsible of the signin and to redirect to the calling application. Once we land back on the calling application from the standalone login page, we trigger the authentication flow using getWithRedirect method to retrieve ID token/ Access token and once those are available, a custom application cookie is created telling that login was successful on one of the three applications A, B or C. 
Now when you open application B, it will check for the existence of this custom cookie, if present, then application B will trigger authentication flow to retrieve ID token/ Access token.

Since we use the tokenManager which stores in the storage the tokens after the authentication flow, the issue occurs when:
- User login from application A => authentication flow triggered => custom cookie created
- User go to application B => custom cookie exists => authentication flow is triggered
- User logout from application A and login with a**nother** user => custom cookie deleted then created again after login with the new user
- User go to application B => since previous tokens are still in the storage and valid AND custom cookie is still here (because it was deleted then added back from application A), then user ends up considered logged in with first user account on application B.

If we have no way to apply the suggested solution by @denysoblohin-okta since it is not compatible with third parties cookies disabled, do you suggest any other options?

Thanks a lot :)

**Comment by jaredperreault-okta** [comment]
Could you store some identifier (like a uuid) in your custom cookie which is unique to the authenticated user, so when you check for the existence of the custom cookie, you can compare that uuid to the value in the stored tokens. If they are not the same, you know the current tokens to not "belong" to the current session (the custom cookie)

**Comment by boubou158** [comment]
Thanks @jaredperreault-okta and @denysoblohin-okta for your hints.

I was able to figure out a solution for this use case considering than when a logout occurs from any application, all custom cookies matching the pattern custom-cookie-*** are cleared.

I created a custom cookie storage provider following the same logic than the SDK (store in two differents cookies id token and access token), only difference with this custom cookie storage provider is that I am able to set the parent domain compared to the native one from the SDK.

Now, ID token and access token are stored in two cookies instead of localStorage, and when a logout occurs from any application, those cookies are cleared. So next time user login from any application, there are no more tokens in the tokenManager and hence the authentication flow will be triggered with the new user.
